2nd Iteration's changes

ahmetkaymet opened the pull request:
- Added detailed database schema documentation to README.md
- Added table and column descriptions for Products, Transactions, and Stores tables
- Added instructions for viewing table descriptions using SQLite commands
- Documented the structure and purpose of table_column_descriptions table
- Removed hardcoded prompts from utils.py and moved to database documentation

OTHER DETAILS WİLL ADD BUT YOU CAN CHECK THE OTHER CHANGES FROM COMMIT DESCRIPTIONS

ahmetkaymet committed:
Update README.md

little fix

Commit:
Update utils.py with debug logs and README updates

ahmetkaymet committed:
.db ignore removed

ahmetkaymet committed:
refactor: remove redundant try-except blocks

ahmetkaymet committed:
security: restrict SQL queries to SELECT only

ahmetkaymet committed:
security: ensure generate_sql_query only produces SELECT queries

ahmetkaymet committed:
-Removed HTTPException import (no longer used)
-Removed try-except blocks (using FastAPI's native error handling)
-Fixed return type for execute-sql endpoint: Dict[str, Any | List[Dict[str, Any]]]
-Implemented cleaner structure using process_natural_query in both endpoints
-Standardized response format:
..generate-sql: {"explanation": str, "query": str}
..execute-sql: {"explanation": str, "results": List[Dict[str, Any]]}
--Code is now cleaner, type-safe, and better aligned with FastAPI best practices.

ahmetkaymet committed:
Refactor: Improve utils.py code structure and maintainability - Enhanced type annotations, simplified docs, optimized code structure

ahmetkaymet committed:
feat: Add detailed comments and improve code documentation in utils.py

- Added descriptive comments for each function
- Improved function documentation with clear explanations
- Added comments explaining database operations
- Clarified OpenAI model interaction process
- Added explanatory comments for SQL query extraction

egemenNB commented on utils.py:
```diff
@@ -18,30 +18,38 @@
 client = OpenAI(api_key=api_key)
 
-
+#connection to database using sqlite library it converts the database to a row factory (dictionary format)
```
You should include the return type of the function in the function's description. The goal of function docstrings are to describe the function task, input & return types, and if present the raised error type and reason.
- You can see this link for Google's Python styling guide with in depth explanations. Feel free to ask questions!
- : https://google.github.io/styleguide/pyguide.html
<img width="806" alt="image" src="https://github.com/user-attachments/assets/469b3940-ed69-4cf9-bdb0-5b0aa541e54a" />

egemenNB commented on utils.py:
````diff
@@ -54,18 +62,19 @@ def generate_sql_query(natural_query: str) -> tuple[str, str]:
     sql_end = full_response.find("```", sql_start + 5)
     sql_query = full_response[sql_start + 6:sql_end].strip()
 
+    #it returns the response and the sql query
````
Let's include the return type in the function docstring

egemenNB reviewed (changes requested):
Good work, let's do a round of review!

ahmetkaymet committed:
docs: improve function documentation according to Google Style Guide

- Remove inline comments and move to docstrings

- Add detailed function behavior explanations

- Document context manager and row_factory usage

- Add markdown parsing details

- Improve docstring formatting

ahmetkaymet commented on utils.py:
```diff
@@ -18,30 +18,38 @@
+#connection to database using sqlite library it converts the database to a row factory (dictionary format)
 def get_db_connection() -> sqlite3.Connection:
     """Create and return database connection"""
+    #it connects to the database
```
I deleted all markdown comments. As you said all codes understandable. In complex lines, I'll use basic markdown comments.
Thanks.

ahmetkaymet commented on utils.py:
```diff
@@ -18,30 +18,38 @@
     """Create and return database connection"""
+    #it connects to the database
     conn = sqlite3.connect("data.db")
+    #it converts the database to a row factory (dictionary format)  
```
I added return types to the docstring. I switched  back to docstring format in comments.

ahmetkaymet committed:
docs: remove unimplemented ValueError from docstring

Remove Raises section from generate_sql_query docstring since ValueError is not actually implemented

ahmetkaymet committed:
docs(api): Enhance API documentation with detailed docstrings

- Add comprehensive module-level documentation with usage example
- Enhance model docstrings with detailed attribute descriptions
- Add request/response examples to endpoint docstrings
- Include data format specifications and type information
- Document the complete request-response flow for each endpoint

ahmetkaymet committed:
refactor(utils): Enhance SQL generation and schema handling

Major improvements in utils.py:

1. Schema Management:
- Enhanced get_db_schema() to include detailed column descriptions
- Added structured formatting for table and column documentation
- Improved schema readability by grouping descriptions by table

2. GPT Prompt Engineering:
- Implemented structured 4-phase approach for SQL generation:
  * ANALYZE: Schema and constraints analysis
  * PLANNING: Query structure and requirements
  * GENERATION: SQL query creation
  * EXPLANATION: Natural language description
- Added strict validation for data formats (e.g., state codes, case sensitivity)
- Enhanced example format with real-world use cases

3. Query Processing:
- Refactored process_natural_query() to return (explanation, sql_query)
- Separated query generation from execution for better modularity
- Improved type hints and documentation

These changes improve query accuracy, maintainability, and error handling
while providing better guidance to the GPT model for SQL generation.

ahmetkaymet committed:
feat: Add query history database and core functionality

- Add query_history.py with core history tracking functions:
  * generate_session_id(): UUID based session tracking
  * save_query_history(): Store queries with results
  * get_session_history(): Retrieve session data
  * initialize_history_db(): Setup database schema

- Add query_history.db schema:
  * id: Auto-incrementing primary key
  * session_id: UUID for query grouping
  * natural_query: Original user query
  * sql_query: Generated SQL
  * gpt_explanation: Detailed analysis
  * query_result: JSON results
  * timestamp: Auto-generated

- Add query_history_schema.sql with:
  * Table creation commands
  * Column descriptions
  * Initial setup queries

ahmetkaymet committed:
docs: Add comprehensive docstrings to query history files

- Add Google-style docstrings to query_history.py
- Add detailed SQL comments to query_history_schema.sql
- Include usage examples and type hints
- Improve documentation clarity and completeness

ahmetkaymet committed:
docs(utils): Enhance process_natural_query documentation - Add detailed step-by-step process explanation - Document query history saving functionality

ahmetkaymet committed:
feat(aimet): Enhance Aimet's personality with background story - Add personal details (age, creator, location) - Include Turkish greeting and heritage - Add ability to discuss personal background

egemenNB commented on utils.py:
```diff
@@ -100,39 +100,62 @@ def generate_sql_query(natural_query: str) -> Tuple[str, str]:
     prompt = f"Database Schema:\n{schema}\n\nUser Query:\n{natural_query}"
 
     system_content = (
```
You can rename this to system_prompt to follow the best practice

ahmetkaymet replied:
Thanks,, I will address that issue. I have also completed all the tasks and will make the commit tomorrow. Additionally, I implemented a basic UI that displays query history along with timestamps to enhance demo testing. Could you let me know if there are any other tasks I should work on?

egemenNB reviewed (commented):
The system prompt is very well structured, and includes in depth information about how the AI should answer. The instructions are clear, and the personality update looks very good as well. You did a great work!

egemenNB commented on utils.py:
```diff
@@ -100,39 +100,62 @@ def generate_sql_query(natural_query: str) -> Tuple[str, str]:
-        "- You MUST mention in your explanation any data format considerations"
+        "Let me explain how this works: [Your explanation]\n\n"
+        
+        "IMPORTANT RULES (you take these very seriously):\n"
```
Let's also ensure that the assistant would never create SQL queries that result in write operations on the db. All queries *must* be strictly read-only.

ahmetkaymet committed:
Improve code clarity and add title generation

- Renamed system_content variable to system_prompt in generate_sql_query function to better reflect its purpose as a system prompt for the GPT model

- Added generate_title function to create concise, descriptive titles (max 50 chars) for queries using GPT-4o

- The title generation helps in better organizing and tracking query history

- Updated related function calls in process_natural_query to include title generation

ahmetkaymet committed:
Enhance API documentation and logging

Major improvements:

1. Added comprehensive docstrings for middleware and logging:

   - Detailed logging configuration documentation

   - CORS middleware purpose and settings

   - Request logging middleware documentation

   - Improved function and class docstrings

2. Code organization and cleanup:

   - Structured logging configuration with clear purpose

   - Better organized middleware section with detailed comments

   - Added request/response logging for debugging

   - Added error tracking and monitoring capabilities

3. Technical details:

   - Configured logging with timestamp, logger name, level, and message

   - Added CORS for frontend communication (ports 3000, 3001)

   - Implemented request-level logging middleware

   - Added error handling with detailed logging

ahmetkaymet committed:
feat: Major enhancements to query history module - Replace get_session_history with get_all_sessions for better session management - Add parse_query_result function for flexible result parsing - Add title field to queries - Improve type hints and documentation - Enhance JSON parsing with fallback mechanism

ahmetkaymet committed:
feat: Add Next.js frontend with comprehensive documentation - Modern UI with Chakra UI components - Real-time query processing and results display - Session management and history tracking - Comprehensive error handling - Detailed documentation and setup instructions

egemenNB commented:
Hi Ahmet! Can you please record a Loom video demonstrating how your application works?
https://www.loom.com/

You can use this chrome extension:
https://chromewebstore.google.com/detail/loom-%E2%80%93-screen-recorder-sc/liecbddmkiiihnedobmlmillhodjkdmb

ahmetkaymet committed:
refactor: migrate from Next.js to Vite + React

ahmetkaymet committed:
feat(frontend): improve query history display - Add timestamp field, show all sessions, improve time display, fix keys

ahmetkaymet committed:
feat(backend): enhance main API endpoints and error handling

ahmetkaymet committed:
feat(backend): improve query history module with better session handling and timestamp support

ahmetkaymet committed:
feat(backend): enhance database utilities and SQL generation

ahmetkaymet commented:
> Hi Ahmet! Can you please record a Loom video demonstrating how your application works? https://www.loom.com/
> 
> You can use this chrome extension: https://chromewebstore.google.com/detail/loom-%E2%80%93-screen-recorder-sc/liecbddmkiiihnedobmlmillhodjkdmb

Sure! I'll share today.

ahmetkaymet committed:
docs: update README with correct port numbers and improved setup instructions - Frontend port changed from 3000 to 5173 (Vite default) - Added detailed troubleshooting section - Updated technology stack (Next.js to Vite) - Added prerequisites section - Improved installation steps

ahmetkaymet committed:
docs: fix terminology in frontend setup instructions

ahmetkaymet committed:
fix: update frontend dependencies - Fix React version to stable 18.2.0 - Remove duplicate dependencies - Properly separate dependencies and devDependencies

egemenNB reviewed (changes requested):
There are 4k+ file changes, so let's add `node_modules` to our  gitignore

ahmetkaymet committed:
Improve Aimet character and frontend enhancements

1. Enhanced system prompt for natural Aimet character representation\n2. Added markdown and code block support to frontend\n3. Fixed proper display of API responses in frontend

ahmetkaymet committed:
Update main.py with improved functionality and bug fixes for the backend API endpoints
